### ISSUE TITLE
something confuse about input

### ISSUE THREAD

**Issue by MRuAyAN** [airalcorn2/Recurrent-Convolutional-Neural-Network-Text-Classifier#7]
1
--
# We shift the document to the right to obtain the left-side contexts.
  | left_context_as_array = np.array([[MAX_TOKENS] + tokens[:-1]])
  | # We shift the document to the left to obtain the right-side contexts.
  | right_context_as_array = np.array([tokens[1:] + [MAX_TOKENS]])

would you mind explain why is this? i think equation should like this:
left_context_as_array = np.array([token[:]])  and 
right_context_as_array = np.array( [token[::-1]] ) .

**Comment by airalcorn2** [comment]
See equations _(1)_, _(2)_, and _(3)_ from the paper. _x<sub>i</sub>_ is defined as the concatenation of _c<sub>l</sub>(w<sub>i</sub>)_ (the left context), _e(w<sub>i</sub>)_ (the word embedding), and _c<sub>r</sub>(w<sub>i</sub>)_ (the right context). _c<sub>l</sub>(w<sub>i</sub>)_ is calculated using _w<sub>i-1</sub>_ while _c<sub>r</sub>(w<sub>i</sub>)_ is calculated using _w<sub>i+1</sub>_. For example,  _x<sub>2</sub>_ (using the paper's numbering scheme) for [the example text in my code](https://github.com/airalcorn2/Recurrent-Convolutional-Neural-Network-Text-Classifier/blob/master/recurrent_convolutional_keras.py#L53) would be the concatenation of the left context calculated for "This" (the first word in the sentence), the embedding for "is" (the second word in the sentence), and the right context calculated for "some" (the third word in the sentence). Because _w<sub>0</sub>_ and _w<sub>n+1</sub>_  are not defined for any given sentence, _c<sub>l</sub>(w<sub>1</sub>)_ and _c<sub>r</sub>(w<sub>n</sub>)_ must be learned separately as discussed in the **Training Network parameters** subsection of the **Training** section. The easiest way I could see to do this in Keras was by defining _w<sub>0</sub>_ and _w<sub>n+1</sub>_ as NULL tokens of zeroes and shifting the input sentences to the right and left, respectively. The reason I don't reverse `right_context_as_array` is because I tell the model to [read it backwards](https://github.com/airalcorn2/Recurrent-Convolutional-Neural-Network-Text-Classifier/blob/master/recurrent_convolutional_keras.py#L39).

**Comment by MRuAyAN** [comment]
I see, you explain so detailed. Thank you so much. A lambda layers will throws some error  when saving this model, have you met it?

**Comment by airalcorn2** [comment]
Can you share the error you're getting?

**Comment by MRuAyAN** [comment]
like this https://github.com/fchollet/keras/issues/6348 and https://github.com/fchollet/keras/issues/3927.
I just use 
pool_rnn = GlobalMaxPooling1D()(semantic) replace 
pool_rnn = Lambda(lambda x: backend.max(x, axis=1), output_shape=(hidden_dim_2,))(semantic)
,thus can save model well.

**Comment by MRuAyAN** [comment]
![tim 20170712164521](https://user-images.githubusercontent.com/18624289/28109507-e3991290-6721-11e7-9346-892e644f965d.png)
Like figure trian precision is okay, but validate precision almost invariant. i have add dropout layer in model. How can i avoid such situation.

**Comment by airalcorn2** [comment]
I do not get that error with Python 3. Sorry, I can't help you tune your model. Looks like you probably need more/better data.
